_posts/2015-09-29-Asking-Users-for-Help.md

```diff
@@ -0,0 +1,155 @@
+
+Recently we had an issue which was pretty hard to figure out. Users started to report that their settings within the app would partly disappear. For some it was a permanent issue and for others it just happened once and then disappeared again. Even a couple of colleagues reported this problem but sadly for us and luckily for them, it disappeared immediately again, so this didn’t help. More and more reports arrived in our mailbox but we were not able to reproduce the issue.
+The fact that we not only store the settings locally but also sync them with our backend added even more complexity to the problem. We checked our crash logs but there was nothing that could explain the problem. We checked with the backend if they had changed anything in the last weeks but they assured us nothing was touched. Plus, all other platforms worked fine so the issue must have been on our side. 
+We started to check our code. First I checked alone but found nothing. Then I asked a colleague to go with me through the logic in case I was just blind. Still we found nothing and still more and more users seemed to encounter the problem.
+This was a programmer's nightmare! A bug keeps getting reported but you don’t have any crash logs and you have no idea how to reproduce the problem.
+
+##The Plan
+
+Giving up was not an option, so we decided to ask those for help, who actually could reproduce the problem. Our users!
+Our plan was to create a special version of the app with extensive logging. We hoped that our users would be kind enough to install this version and by doing so, help us identify the issue.
+
+##Have theories you want to verify
+
+We had several suspicions / theories which we wanted to check:
+
+- Race conditions in our cache. 
+ - <i>This is a problem that typically occurs when you have multiple threads depending on a shared state and you failed to synchronize the different threads properly. The evil thing about this issue is, that it heavily depends on the execution order of your threads. Everything can go fine a hundred times but every once in awhile your threads randomly get processor time in a specific order which causes the bug to appear. And when you have several million users this will happen a lot even if it does not happen for you!</i>
+
+- A logic problem which we just did not see when checking the code.
+
+- Some faulty error handling while syncing with the backend.
+
+- A server instance sending us corrupt data for some reason.
+
+In general it seems to be helpful if you have at least some idea where you have vulnerable parts in your code. This makes it much easier to decide what to log.
+
+##What do you want to log?
+
+To be able to properly judge which theories if any are true and wich are faulty we decided to log the following data:
+
+- user id
+
+- exact time you log something
+- The overall uptime of the device when you log something
+ - This helps you to detect restarts of the device
+- The screens the user is navigating to
```

Comment:
To be more consistent:
  - The user ID
  - The exact time the log is made
  - The overall uptime of the device at that time
    - This helps you to detect restarts of the device